### ISSUE TITLE
CONFIG_X86_P6_NOP is unset outside MPENTIUM4 || MCORE2 || MPSC

### ISSUE THREAD

**Issue by tpruzina** [graysky2/kernel_gcc_patch#19]
On core2+ choices CONFIG_X86_P6_NOP= is set to N as it depends on ``X86_64 [=y] && (MCORE2 [=n] || MPENTIUM4 [=n] || MPSC [=n])``.

Fix seems fairly straightforward unless user sets march=native option, perhaps it would be worth to modify patch and make this option user specifiable if march native is set (assuming non intel hardware or < P6).

``arch/x86/Kconfig.cpu``
<pre>
# P6_NOPs are a relatively minor optimization that require a family >=
# 6 processor, except that it is broken on certain VIA chips.
# Furthermore, AMD chips prefer a totally different sequence of NOPs
# (which work on all CPUs).  In addition, it looks like Virtual PC
# does not understand them.
#
# As a result, disallow these if we're not compiling for X86_64 (these
# NOPs do work on all x86-64 capable chips); the list of processors in
# the right-hand clause are the cores that benefit from this optimization.
#
config X86_P6_NOP
        def_bool y
        depends on X86_64
        depends on (MCORE2 || MPENTIUM4 || MPSC)
</pre>

**Comment by graysky2** [comment]
PRs welcomed :)

**Comment by tpruzina** [comment]
PR would require some manual hackery (e.g. patching a patch) unfortunatedly, but I fixed it like this on my machine:

https://gist.github.com/tpruzina/8ecf390b744925adf1f4cd32c51f2300#file-p6_nop-patch

Note that it's not the prettiest option out there to fix this but it doesn't break existing configs.
1) non-intel CPU selected - option defaults to N and stays hidden
2) intel CPU selected - option is automatically selected
3) march=native - option is shows up, initially set to N (so it doesn't break on non-intel cpus)

**Comment by graysky2** [comment]
I will review and incorporate.  I think your proposal fixes cases 1 and 2 you outlined above.  I am not sure would case 3 though as you point out... perhaps it's best to remove the native option.

**Comment by graysky2** [comment]
Have a look at the two most recent commits in the dev branch.

**Comment by eli-schwartz** [comment]
Building [linux-ck](https://aur.archlinux.org/packages/linux-ck) for i686 and selecting MCORE2, I get this warning repeated a lot:

`warning: (MATOM && MCORE2 && MNEHALEM && MWESTMERE && MSILVERMONT && MSANDYBRIDGE && MIVYBRIDGE && MHASWELL && MBROADWELL && MSKYLAKE) selects X86_P6_NOP which has unmet direct dependencies (X86_64 && (MCORE2 || MPENTIUM4 || MPSC || MATOM || MNEHALEM || MWESTMERE || MSILVERMONT || MSANDYBRIDGE || MIVYBRIDGE || MHASWELL || MBROADWELL || MSKYLAKE || MNATIVE))`

**Comment by graysky2** [comment]
Will look into this...
@tpruzina - Any thoughts?

**Comment by tpruzina** [comment]
Hm, warning seems to be safe to ignore, will take a look into this as x86_64 _capable_ hardware _should_ be supported, even on i686.

**Comment by graysky2** [comment]
> It would seem that current dependence on x86_64 was included as a workaround for buggy microsoft hypervisor host. I feel pretty confident on removing this dependency.

@eli-schwartz - Would you mind testing this hypothesis?

**Comment by eli-schwartz** [comment]
I reversed the previous patch and reapplied this one, then re-ran the kernel config. It generated the exact same .config but without the warning.

(Well, **technically**, the new .config had a couple new commented-out options, for excavator and zen.)

**Comment by graysky2** [comment]
@eli-schwartz - I just applied this version to linux from ABS and build in a 32-bit chroot without errors.  Guess I'll merge into master and call this good.

**Comment by tpruzina** [comment]
@eli-schwartz @graysky2 thanks
